### PR TITLE
Add upsert option for insert call. Only works for PostgreSQL >=9.5

### PR DESCRIPTION
**Pull request by gotar**
using Sequel native support

### CODE REVIEW

**Comment by gotar** [comment]
@solnic WDYT? I was thinking a lot, in which direction it should go and how to implement this, but this one win. But I'm open to any suggestions? Not sure you want sth like this?

**Comment by solnic** [comment]
I updated this branch and added an explicit `Postgres::Upsert` command class. This will allow us to fine-tune `upsert` behavior per-command basis.

I think this needs a bit more work so that it's configurable. Personally I'd like to achieve a smart `update_or_delete` behavior. It seems like it's possible via `dataset.insert_conflict(custom_opts).insert(tuple)`.

@flash-gordon do you have any experience with this?

**Comment by flash-gordon** [comment]
@solnic wdym by `update_or_delete`? I don't get you :)

**Comment by solnic** [comment]
ugh, i'm tired...I meant `create or update`

**Comment by flash-gordon** [comment]
That's what I thought, just decided to make it clear 😁  I haven't worked with PG's upserts yet, only with Oracle's `merge` operator which is waaaay more complex (and not atomic in contrast).

Anyway, according to the [docs](https://www.postgresql.org/docs/9.5/static/sql-insert.html) there are following options for `insert_on_confict`:
1. Action. It can be `do nothing` or `do update`. I think it is not needed to be set explicitly, the fact that we have (not) an update statement is sufficient for choosing a proper option.
2. Constraint name. I think this _should_ be configurable, though this is not a mandatory parameter.
3. Conflict target. This is an optional set of indexed columns that can be set _instead of_ constraint name. I'm not sure why and when this option should be used, but I guess this can be helpful when constraint name is unknown, ie auto-generated or sutin. We _could_ support for this option, the corresponding Sequel's parameter named `:target`.
4. Update. This is a little bit tricky. You can provide a static value for it, this means you'll be able to update an existing row having access to both existing row and so called _excluded_ row in SQL. It looks like you provide a transformation for the existing row expressed via a small piece of SQL. You can also provide this update statement for every `upsert` call be I doubt you'll ever need to do so. I think static configuration is good enough, and we _should_ allow to configure it.
5. Where clause for update part. We _could_ support for this but this would be used very occasionally fwiw. 

Sequel's docs http://sequel.jeremyevans.net/rdoc-adapters/classes/Sequel/Postgres/DatasetMethods.html#method-i-insert_conflict

File: lib/rom/sql/commands/postgres.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+module ROM
+  module SQL
+    module Commands
+      module Postgres
```
Comment (flash-gordon) [inline lib/rom/sql/commands/postgres.rb]: I believe this should be named PG to be consistent with ROM::SQL::Types::PG

Comment (flash-gordon) [inline lib/rom/sql/commands/postgres.rb]: Though the next file named `lib/rom/sql/commands_ext/postgres.rb` :)

**Comment by flash-gordon** [comment]
I added support for all Sequel's (ie PG's) options you can provide for `insert on conflict` statement. Though I'm not confident with the API, I added specs for these options and everything works smooth so far.
/cc @gotar @solnic

**Comment by flash-gordon** [comment]
I fixed specs for `Create` command (oops). It's worth mentioning that PG does not return anything on `ON CONFLICT DO NOTHING` and we can't find out what record exactly rejected the insert. I have no idea how to handle this case, probably the end user will need to update some dummy column to force PG to return the conflicted row.

**Comment by flash-gordon** [comment]
Now when the specs use `insert on conflict` they are failing on Travis because it uses PG 9.1 by default. Using 9.5 on Travis is not so easy https://github.com/travis-ci/travis-ci/issues/4264, we either should not run that specs on PG < 9.5 or run them against 9.5 on Travis. The later option could slow down the build significantly so this requires some research beforehand, I'm going to do it today or tomorrow.

**Comment by solnic** [comment]
@flash-gordon circle ci supports PG 9.5 OOTB maybe we should move there

**Comment by flash-gordon** [comment]
@solnic not a bad idea, I use circle ci and I can say it is nice. I'll try to setup the environment for my fork first

**Comment by solnic** [comment]
@flash-gordon yeah I've been using CircleCI a lot, it's super cool. Setup should be straight-forward. IIRC the provide parallel builds for OSS projects too, which could significantly speed up our builds if we split specs per adapter :) I enhanced our setup with ability to run specs for a specific adapter on-demand, it's not controlled by external variables yet though, but this could be easily done so that we could do stuff like `ADAPTER=sqlite rake spec`

**Comment by gotar** [comment]
sequel support (except pgsql) mysql and sqlite if I good remember. So it should be very easy to extend this functionality to those databases too.

File: spec/integration/commands/upsert_spec.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        expect(command.call(excluded)).to be nil
+      end
+    end
+
+    context 'on conflict do update' do
+      context 'with conflict target' do
+        let(:command_config) do
+          -> do
+            conflict_target :title
+            update_statement user_id: 2
```
Comment (solnic) [inline spec/integration/commands/upsert_spec.rb]: Does it mean it can't figure out the update_statement at runtime?

Comment (flash-gordon) [inline spec/integration/commands/upsert_spec.rb]: @solnic the only reason why this is not possible is that command interface is lack of a method that builds a new command with new options
```ruby
          command_with_update = command.with_options(
            update_statement: {
              user_id: :excluded__user_id
            }
          )
```
But anyway, you are veeery unlikely need to set update_statement at runtime, that's why this is a class level option by default.

**Comment by solnic** [comment]
@flash-gordon anything else left to do here?

**Comment by solnic** [comment]
It looks good and it's a great starting point, I suspect we're gonna need to make some improvements to make it work smoothly with repos, but that can be done separately.

Thanks guys :)

**Pull request merged**